packages/contracts/contracts/zkopru/controllers/Migratable.sol

```diff
@@ -1,84 +1,127 @@
-        require(chain.migrations[migrationId], "MassMigration does not exist");
-        try
-            Migratable(to).acceptMigration(
-                migrationId,
-                migration.migratingLeaves.merged,
-                migration.migratingLeaves.fee
-            )
-        {
-            // TODO: Handle out of gas due to the push pattern => ex: slash proposer using checksum?
-            // send ETH first
-            payable(to).transfer(migration.totalETH);
-            // send ERC20
-            for (uint256 i = 0; i < migration.erc20.length; i++) {
-                IERC20(migration.erc20[i].addr).transfer(
-                    to,
-                    migration.erc20[i].amount
-                );
-            }
-            // send ERC721
-            for (uint256 i = 0; i < migration.erc721.length; i++) {
-                for (uint256 j = 0; j < migration.erc721[i].nfts.length; j++) {
-                    IERC721(migration.erc721[i].addr).transferFrom(
-                        address(this),
-                        to,
-                        migration.erc721[i].nfts[j]
-                    );
-                }
-            }
-            // Delete mass migration
-            delete chain.migrations[migrationId];
-        } catch {
-            revert("Dest contract denied migration");
+        MassDeposit memory massDeposit = migration.depositForDest;
+        address token = migration.asset.token;
+        // 1. fetch prev balances
+        uint256 prevEthBal = address(this).balance;
+        uint256 prevTokenBal;
+        if (token != address(0)) {
+            prevTokenBal = IERC20(token).balanceOf(address(this));
+        }
+        // 2. transfer migration asset
+        Migratable(source).transfer(migrationRoot, migration, index, siblings);
+        // 3. fetch new balances
+        uint256 newEthBal = address(this).balance;
+        uint256 newTokenBal;
+        if (token != address(0)) {
+            newTokenBal = IERC20(token).balanceOf(address(this));
+        }
+        // 4.ccheck asset transfer validity
+        require(
+            prevEthBal.add(migration.asset.eth) == newEthBal,
+            "Invalid ETH transfer"
+        );
+        require(
+            prevTokenBal.add(migration.asset.amount) == newTokenBal,
+            "Invalid Token transfer"
+        );
+        //  5. Check Mass Deposit Leaves
+        bytes32 merged;
+        for (uint256 i = 0; i < leaves.length; i++) {
+            merged = keccak256(abi.encodePacked(merged, leaves[i]));
         }
+        require(
+            merged == migration.depositForDest.merged,
+            "Data Availability problem."
+        );
+        // 6. Record the carried mass deposit as committed
+        bytes32 migrationHash = migration.hash();
+        bytes32 depositHash = massDeposit.hash();
+        Storage.chain.committedDeposits[depositHash] += 1;
+        emit NewMassMigration(source, migrationRoot, migrationHash);
```

Comment:
This is a more general question, but can anyone trigger a mass migration? If I'm an attacker could I deploy a modified Zkopru contract and call `migrateFrom` with my contract that assumes control of all assets as the destination?

Reply:
Yes the deployed contract can act as a destination with its contract address. But to execute the `transfer()` in the line 50, the migration should be included in the finalized migration root of the source network (line 87 & 99).

Reply:
How does a migration get included in the migration root of the source network? I've always been unclear on this mechanism. Do users individually include their assets in a migration e.g. I can move all my Ether/ERC20's to the new contract but not other peoples?

Reply:
Let's assume 
* Alice has three notes and wants to migrate them from A to B & C.
    * a1: ETH note (A=>B)
    * a2: ETH note (A=>B)
    * a3: DAI note (A=>C)

* Bob has two notes and
    * b1: ETH note (A=>B)

* Carl has 2 note for DAI
    * c1: DAI note (A=>B)
    * c2: DAI note (A=>C)
***
Tx gen step
* Alice sends a tx that consumes 3 notes and generates 3 migration leaves.
* Bob sends a tx that consumes 1 note and generates 1 migration leaf.
* Carl sends a tx that consumes 2 notes and generates 2 migration leaves.
***
Block gen step
* Then the coordinator will include those 3 transactions and will find out that there exist migration leaves and their destinations are "B" and "C".
* For destination "B", there exist ETH notes & DAI notes. Otherwise, for destination "C" there only exists "DAI" notes.
* So, the coordinator creates 3 MassMigration objects that 
    * M1: ETH migration for dest "B" ([a1, a2, b1])
    * M2: DAI migration for dest "B" ([c1])
    * M3: DAI migration for dest "C" ([c2])
* And include them all into the block.
* If there existed a transaction in the proposed block that is migrating a leaf to destination "D" and the coordinator emits to include the MassMigation for that leaf, it'll get slashed :dagger:

Reply:
So in the UI we'll need each user to submit a migration transaction for their assets?

Reply:
yes right, so if we have an upgraded zkopru network, we can show the option for migration to the wallet users.